aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1658,101 +1666,120 @@ Tensor _matmul_impl(
+                      : MaybeOwned<Tensor>::borrowed(transpose ? tensor1 : tensor2);
+    // Invariant: t1->dim() >= 3 && (t2->dim() == 1 || t2->dim() == 2)
+
+    // Why not t1->view(-1, sizes_1.back())?
     // If the last dim is 0, then view(-1, 0) won't work because the -1 becomes ambiguous.
     // This can happen in e.g. [3, 5, 0] @ [0, 0].
-    // So we manually compute the folding as a result.
-    const auto dim1_size = c10::multiply_integers(size1.begin(), size1.end() - 1);
-    auto t1 = tensor1.expect_contiguous()->view({dim1_size, size1[size1.size() - 1]});
-    Tensor output = has_out ? at::_unsafe_view(at::mm_out(out, t1, t2), output_size)
-                            : at::_unsafe_view(t1.mm(t2), output_size);
-    return has_out ? out.set_(output) : output;
-  } else if ((dim_tensor1 == 1 || dim_tensor1 == 2) && dim_tensor2 >= 3) {
-    // optimization: transpose the inner dimensions of the arguments, call
-    // matmul on the swapped arguments, then transpose the inner dimensions
-    // of the result.
-    const int64_t n = dim_tensor1 == 2 ? tensor1.size(-2) : 1;
-    const int64_t m = tensor1.size(-1);
-    const int64_t p = tensor2.size(-1);
-
-    const Tensor t2_T = tensor2.transpose(-1, -2);
-    const Tensor t1_T = dim_tensor1 == 2 ? tensor1.t() : tensor1.reshape({n, m}).t();
-    const Tensor res_T = _matmul_impl(out, t2_T, t1_T);
-
-    if (dim_tensor1 == 2) {
-      Tensor res = res_T.transpose(-1, -2).contiguous();
-      return has_out ? out.set_(res) : res;
+    const auto sizes_1 = t1->sizes();
+    auto output_shape = DimVector(sizes_1.begin(), sizes_1.end() - 1);
+    const auto batch_product = c10::multiply_integers(output_shape);
+
+    // Readjust output_shape if we are multiplying by a matrix
+    const auto t2_is_matrix = t2->dim() == 2;
+    if (t2_is_matrix) {
+      output_shape.push_back(t2->sizes()[1]);
     }
-    else {
-      c10::IntArrayRef shape_array = tensor2.sizes().slice(0, dim_tensor2 - 2);
-      DimVector shape(shape_array.begin(), shape_array.end());
-      shape.push_back(p);
+    const auto t1_folded = t1->reshape({batch_product, sizes_1.back()});
+    if (!has_out) {
+      if (t2_is_matrix) {
+        // FIXME This path always does an unnecessary copy when transpose == true as the returned
+        // result from BLAS is already C-transposed
+        const auto output = at::_unsafe_view(t1_folded.mm(*t2), output_shape);
+        return transpose ? output.transpose_(-2, -1).contiguous() : output;
+      } else {
+        return at::_unsafe_view(t1_folded.mv(*t2), output_shape);
+      }
+    } else {
+      // Resize output into the correct shape
+      const auto transpose_out = transpose && t2_is_matrix;
+      if (transpose_out) {
+        // Swap last two elements of output_shape
+        std::iter_swap(output_shape.end() - 2, output_shape.end() - 1);
+        at::native::resize_output(out, output_shape);
+        out.transpose_(-2, -1);
```

Comment:
This looks sketchy, we're not supposed to edit the metadata for a user given out tensor, we have to respect what the user asked us to write into (unless a resize occurred).

Reply:
Discussing this here https://github.com/pytorch/pytorch/pull/75197#discussion_r843896392 resolving this one.